src/v/storage/disk_log_impl.cc

```diff
@@ -534,8 +534,13 @@ ss::future<compaction_result> disk_log_impl::compact_adjacent_segments(
             throw std::runtime_error(fmt::format(
               "Aborting compaction of closed segment: {}", *segment));
         }
+        // find max generation of compacted segments
+        result_generation_id = std::max(result_generation_id, *gen_it);
         ++gen_it;
     }
+    // updated and advance resulting segment generation
+    replacement->set_generation(result_generation_id);
```

Comment:
I wonder if we could do this in `make_concatenated_segment`. `self_compact_segment` advances the generation too so we could remove lines 541-543 and have just one code path for dealing with generations instead of two.

Reply:
a partially related question: other than preventing compaction from taking affect, is it always safe to advance generations? meaning, if we accidentally forgot to advance it in some places this could affect correctness, but if we advanced it in too many places then we'd still be correct?

Reply:
advancing generation in to many places is correct, it would slow down compaction/truncation but would not affect correctness